src/libopensc/card-eoi.c

```diff
@@ -0,0 +1,547 @@
+	sbuf[0] = 0x91;
+	sbuf[1] = 0x02;
+	if (env->algorithm_flags & SC_ALGORITHM_ECDSA_HASH_SHA1)
+		sbuf[2] = 0x11;
+	else if (env->algorithm_flags & SC_ALGORITHM_ECDSA_HASH_SHA256)
+		sbuf[2] = 0x21;
+	else if (env->algorithm_flags & (SC_ALGORITHM_ECDSA_RAW|SC_ALGORITHM_ECDSA_HASH_NONE))
+		sbuf[2] = 0x22;
+	else
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INVALID_ARGUMENTS);
+	for (i = 0; i < MAX_OBJECTS && privdata->prkey_mappings[i][1]; i++) {
+		if (privdata->prkey_mappings[i][0] == env->key_ref[0])
+			break;
+	}
+	if (i == MAX_OBJECTS)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INVALID_ARGUMENTS);
+	sbuf[3] = privdata->prkey_mappings[i][1];
+	apdu.lc = 4;
+	apdu.datalen = 4;
+	apdu.data = sbuf;
+	if (se_num > 0) {
+		r = sc_lock(card);
+		LOG_TEST_RET(card->ctx, r, "sc_lock() failed");
+		locked = 1;
+	}
+	if (apdu.datalen) {
+		r = sc_transmit_apdu(card, &apdu);
+		if (r) {
+			sc_log(card->ctx, "%s: APDU transmit failed", sc_strerror(r));
+			goto err;
+		}
+		r = sc_check_sw(card, apdu.sw1, apdu.sw2);
+		if (r) {
+			sc_log(card->ctx, "%s: Card returned error", sc_strerror(r));
+			goto err;
+		}
+	}
+	if (se_num <= 0) {
+		r = SC_SUCCESS;
+		goto err;
+	}
+	sc_unlock(card);
+	LOG_TEST_RET(card->ctx, r, "APDU transmit failed");
+
+	LOG_FUNC_RETURN(card->ctx, sc_check_sw(card, apdu.sw1, apdu.sw2));
+err:
+	if (locked)
+		sc_unlock(card);
+	LOG_FUNC_RETURN(card->ctx, r);
+}
+
+static int eoi_compute_signature(struct sc_card *card, const u8 * data, size_t data_len, u8 *out, size_t outlen)
+{
+	struct eoi_privdata *privdata = (struct eoi_privdata *)card->drv_data;
+	int r;
+
+	LOG_FUNC_CALLED(card->ctx);
+
+	if (!privdata)
+		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INTERNAL);
+
+	/*
+	 * Guess the correct mode. If the size is less than the full-key-len, it must be a hash then
+	 */
+	if (privdata->key_len != data_len) {
+		switch (data_len) {
+			case SHA_DIGEST_LENGTH:
+				privdata->sec_env->algorithm_flags = SC_ALGORITHM_ECDSA_HASH_SHA1;
+				break;
+			case SHA256_DIGEST_LENGTH:
+				privdata->sec_env->algorithm_flags = SC_ALGORITHM_ECDSA_HASH_SHA256;
+				break;
+		}
+	}
+	/* Now we know which hash is used */
+	privdata->sec_env->algorithm_flags |= ALREADY_PROCESSED;
+
+	/* Perform the true set_security_env */
+	r = eoi_set_security_env(card, privdata->sec_env, privdata->se_num);
+	LOG_TEST_RET(card->ctx, r, "set_security_env failed");
+
+	LOG_FUNC_RETURN(card->ctx, sc_get_iso7816_driver()->ops->compute_signature(card, data, data_len, out, outlen));
+}
+
+struct sc_card_driver *sc_get_eoi_driver(void)
+{
+	eoi_ops = *sc_get_iso7816_driver()->ops;
+
+	eoi_ops.match_card = eoi_match_card;
+	eoi_ops.init = eoi_init;
+	eoi_ops.finish = eoi_finish;
+	eoi_ops.select_file = eoi_select_file;
+	eoi_ops.logout = eoi_logout;
+	eoi_ops.pin_cmd = eoi_pin_cmd;
+	eoi_ops.set_security_env = eoi_set_security_env;
+	eoi_ops.compute_signature = eoi_compute_signature;
+
+	return &eoi_drv;
+}
+
+#else
+
+#include "opensc.h"
```

Comment:
can we have this before the `ifdefs` so it is not duplicated and placed awkwardly on the end of the file?